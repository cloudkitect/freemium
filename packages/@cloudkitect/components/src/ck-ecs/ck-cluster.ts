/*
 * Copyright (c) CloudKitect Inc. All Rights Reserved.
 */

import * as ecs from 'aws-cdk-lib/aws-ecs';
import { Construct } from 'constructs';
import { CkVendorTags } from '../ck-core/ck-vendor-tags';
import { CkVpc } from '../ck-ec2/ck-vpc';
import { CkStack } from '../ck-stack/ck-stack';

/**
 * CloudKitect Cluster component properties
 */
export interface CkClusterProps extends ecs.ClusterProps {}

/**
 * CloudKitect ECS cluster component is a regional grouping of one or more container instances on
 * which you can run tasks and services and enables container insights,
 *
 * ### Default Configuration
 * New VPC is created by default.
 *
 * ### Default Alarms
 * Available only in Enhanced components
 *
 * ### Logging and Monitoring
 * Available only in Enhanced components
 *
 * ### Examples
 * Default Usage
 * ```ts
 * new CkCluster(this, "LogicalId", {});
 * ```
 *
 * Custom Configuration
 * ```ts
 * new CkCluster(this, "LogicalId", {
 *    containerInsights: false
 * });
 * ```
 *
 * ### Compliance
 * It addresses the following compliance requirements
 *
 */
export class CkCluster extends ecs.Cluster {
  constructor(scope: Construct, id: string, props: CkClusterProps) {
    super(scope, id, {
      ...props,
      // CC-80: ECS Cluster should allow container insights by default for production
      containerInsights:
                props.containerInsights ?? CkStack.findStackOf(scope).isProduction(),

      // CC-81: VPC
      vpc: props.vpc ?? new CkVpc(scope, 'ClusterVPC', {}),
    });

    CkVendorTags.addTagsTo(this);
  }
}
