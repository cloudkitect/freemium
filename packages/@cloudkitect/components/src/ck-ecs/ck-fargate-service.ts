/*
 * Copyright (c) CloudKitect Inc. All Rights Reserved.
 */

import * as ec2 from 'aws-cdk-lib/aws-ec2';
import * as ecs from 'aws-cdk-lib/aws-ecs';
import { Construct } from 'constructs';
import { CkVendorTags } from '../ck-core/ck-vendor-tags';
import { CkStack } from '../ck-stack/ck-stack';

/**
 * CloudKitect Fargate Service component properties
 */
export interface CkFargateServiceProps extends ecs.FargateServiceProps {}

/**
 * CloudKitect Fargate Service component to run tasks
 *
 * ### Default Configuration
 * VPC Subnet: Private
 *
 * ### Default Alarms
 * Available only in Enhanced components
 *
 * ### Logging and Monitoring
 * Available only in Enhanced components
 *
 * ### Examples
 * ```ts
 * new CkFargateService(this, "LogicalId", {});
 * ```
 *
 * Custom Configuration
 * ```ts
 * new CkFargateService(this, "LogicalId", {
 *    assignPublicIp: true
 * });
 * ```
 *
 * ### Compliance
 * It addresses the following compliance requirements
 * 1. Do not assign public IP
 *    >  - Risk Level: Medium
 *    >  - Compliance: NIST4
 *    >  - Well Architected Pillar:  Security
 */
export class CkFargateService extends ecs.FargateService {

  constructor(scope: Construct, id: string, props: CkFargateServiceProps) {
    super(scope, id, {
      ...props,

      // CC-85: Do not assign public IP address to Fargate Service
      assignPublicIp: props.assignPublicIp ?? false,

      // CC-89: Fargate Service should have circuit breaker with rollback failed deployments
      circuitBreaker: props.circuitBreaker ?? { rollback: true },

      // CC-94: vpc subnet logic
      vpcSubnets: props.vpcSubnets ?? {
        subnetType: ec2.SubnetType.PRIVATE_WITH_EGRESS,
      },
    });

    this.applyRemovalPolicy(CkStack.findStackOf(this).removalPolicy());

    CkVendorTags.addTagsTo(this);
  }

}
