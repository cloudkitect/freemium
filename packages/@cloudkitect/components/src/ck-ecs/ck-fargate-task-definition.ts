/*
 * Copyright (c) CloudKitect Inc. All Rights Reserved.
 */

import * as ecs from 'aws-cdk-lib/aws-ecs';
import { Construct } from 'constructs';
import { CkVendorTags } from '../ck-core/ck-vendor-tags';
import { CkStack } from '../ck-stack/ck-stack';

/**
 * Well Architected Fargate Task Definition properties
 */
export interface CkFargateTaskDefinitionProps
  extends ecs.FargateTaskDefinitionProps {}

/**
 * Well Architected Fargate Task Definition component
 *
 * ### Default Configuration
 * CPU: 2048 (2vCPU) for Production
 * Memory: 4096 GB for Production
 * Storage: 50 GB for Production
 *
 * ### Default Alarms
 * None
 *
 * Note that the default alarm uses the CcAlarm construct, which sets up an alarm
 * action to notify the SNS Topic *AlarmEventsTopic* by default.
 *
 * ### Examples
 * Default Usage
 * ```ts
 * new CcFargateTaskDefinition(this, "LogicalId", {});
 * ```
 *
 * Custom Configuration
 * ```ts
 * new CcFargateTaskDefinition(this, "LogicalId", {
 *    cpu: 2048
 * });
 * ```
 *
 * ### Compliance
 * It addresses the following compliance requirements
 * 1. Encrypted storage
 *    >  - Risk Level: High
 *    >  - Compliance: NIST4
 *    >  - Well Architected Pillar: Security
 * 2. Cost Optimization
 *    >  - Risk Level: Low
 *    >  - Compliance: NA
 *    >  - Well Architected Pillar: Cost Optimization
 *
 */
export class CkFargateTaskDefinition extends ecs.FargateTaskDefinition {
  constructor(
    scope: Construct,
    id: string,
    props: CkFargateTaskDefinitionProps,
  ) {
    super(scope, id, {
      ...props,

      // CC-82: FargateTaskDefinition should have environment dependent cpu capacity
      cpu: props.cpu ?? (CkStack.findStackOf(scope).isProduction() ? 2048 : 256),

      // CC-83: FargateTaskDefinition should have environment dependent memory capacity
      memoryLimitMiB:
                props.memoryLimitMiB ??
                (CkStack.findStackOf(scope).isProduction() ? 4096 : 512),

      // CC-84: FargateTaskDefinition should have environment dependent ephemeral storage capacity
      ephemeralStorageGiB:
                props.ephemeralStorageGiB ??
                (CkStack.findStackOf(scope).isProduction() ? 50 : 21),
    });

    this.applyRemovalPolicy(CkStack.findStackOf(this).removalPolicy());

    CkVendorTags.addTagsTo(this);
  }
}
