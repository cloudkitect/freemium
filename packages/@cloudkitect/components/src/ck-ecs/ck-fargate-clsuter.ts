/*
 * Copyright (c) CloudKitect Inc. All Rights Reserved.
 */

import { Construct } from 'constructs';
import { CkCluster, CkClusterProps } from './ck-cluster';
import { CkVendorTags } from '../ck-core/ck-vendor-tags';
import { CkStack } from '../ck-stack/ck-stack';

/**
 * CloudKitect Fargate Cluster component properties
 */
export interface CkFargateClusterProps extends CkClusterProps {}

/**
 * CloudKitect ECS cluster component backed by Fargate capacity provider is a regional grouping of
 * one or more container instances on which you can run tasks and services and enables container insights,
 *
 * ### Default Alarms
 *
 * ### Examples
 * Default Usage
 * ```ts
 * new CcFargateCluster(this, "LogicalId", {});
 * ```
 *
 * Custom Configuration
 * ```ts
 * new CcFargateCluster(this, "LogicalId", {
 *    containerInsights: false
 * });
 * ```
 *
 * ### Compliance
 * It addresses the following compliance requirements
 * 1. Cluster insights enabled
 *    >  - Risk Level: Medium
 *    >  - Compliance: NA
 *    >  - Well Architected Pillar: Operational Excellence
 */
export class CkFargateCluster extends CkCluster {
  constructor(scope: Construct, id: string, props: CkFargateClusterProps) {
    super(scope, id, {
      ...props,
      // CC-322: Fargate capacity provider
      enableFargateCapacityProviders: true,
    });

    this.applyRemovalPolicy(CkStack.findStackOf(this).removalPolicy());

    CkVendorTags.addTagsTo(this);
  }
}
