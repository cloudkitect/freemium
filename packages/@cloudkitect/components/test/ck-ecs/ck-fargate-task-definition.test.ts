import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { CkAccountType, CkStack } from '../../lib';
import { CkFargateTaskDefinition } from '../../src/ck-ecs/ck-fargate-task-definition';

let app: App;
let stack: CkStack;
const stackProps = {
  ckAccountType: CkAccountType.PRODUCTION,
  ckApplication: 'Test',
  ckCompany: 'CloudKitect',
};

describe('Well Architected ECS Fargate Task Definition', () => {
  beforeEach(() => {
    app = new App();
    stack = new CkStack(app, 'TestStack', { ...stackProps });
  });

  test('default fargate task definition', () => {
    new CkFargateTaskDefinition(stack, 'FGTaskDef', {});

    const template = Template.fromStack(stack);
    expect(template.toJSON()).toMatchSnapshot();
  });

  test.each`
      actual       | accountType                  | expected
      ${undefined} | ${CkAccountType.PRODUCTION}  | ${'2048'}
      ${null}      | ${CkAccountType.PRODUCTION}  | ${'2048'}
      ${'512'}     | ${CkAccountType.PRODUCTION}  | ${'512'}
      ${undefined} | ${CkAccountType.DEVELOPMENT} | ${'256'}
      ${null}      | ${CkAccountType.DEVELOPMENT} | ${'256'}
      ${'512'}     | ${CkAccountType.DEVELOPMENT} | ${'512'}
   `(
    'CC-82: task definition cpu should be $expected when $actual in $accountType',
    ({ actual, accountType, expected }) => {
      const myStack = new CkStack(app, 'MyStack', {
        ...stackProps,
        ckAccountType: accountType,
      });
      new CkFargateTaskDefinition(myStack, 'FGTaskDef', {
        cpu: actual,
      });

      const template = Template.fromStack(myStack);
      template.hasResourceProperties('AWS::ECS::TaskDefinition', {
        Cpu: expected,
      });
    },
  );

  test.each`
      actual       | accountType                  | expected
      ${undefined} | ${CkAccountType.PRODUCTION}  | ${'4096'}
      ${null}      | ${CkAccountType.PRODUCTION}  | ${'4096'}
      ${'512'}     | ${CkAccountType.PRODUCTION}  | ${'512'}
      ${undefined} | ${CkAccountType.DEVELOPMENT} | ${'512'}
      ${null}      | ${CkAccountType.DEVELOPMENT} | ${'512'}
      ${'1024'}    | ${CkAccountType.DEVELOPMENT} | ${'1024'}
   `(
    'CC-83: task definition memory should be $expected when $actual in $accountType',
    ({ actual, accountType, expected }) => {
      const myStack = new CkStack(app, 'MyStack', {
        ...stackProps,
        ckAccountType: accountType,
      });
      new CkFargateTaskDefinition(myStack, 'FGTaskDef', {
        memoryLimitMiB: actual,
      });

      const template = Template.fromStack(myStack);
      template.hasResourceProperties('AWS::ECS::TaskDefinition', {
        Memory: expected,
      });
    },
  );

  test.each`
      actual       | accountType                  | expected
      ${undefined} | ${CkAccountType.PRODUCTION}  | ${50}
      ${null}      | ${CkAccountType.PRODUCTION}  | ${50}
      ${100}       | ${CkAccountType.PRODUCTION}  | ${100}
      ${undefined} | ${CkAccountType.DEVELOPMENT} | ${21}
      ${null}      | ${CkAccountType.DEVELOPMENT} | ${21}
      ${50}        | ${CkAccountType.DEVELOPMENT} | ${50}
   `(
    'CC-84: task definition ephemeral storage should be $expected when $actual in $accountType',
    ({ actual, accountType, expected }) => {
      const myStack = new CkStack(app, 'MyStack', {
        ...stackProps,
        ckAccountType: accountType,
      });
      new CkFargateTaskDefinition(myStack, 'FGTaskDef', {
        ephemeralStorageGiB: actual,
      });

      const template = Template.fromStack(myStack);
      template.hasResourceProperties('AWS::ECS::TaskDefinition', {
        EphemeralStorage: {
          SizeInGiB: expected,
        },
      });
    },
  );
});
