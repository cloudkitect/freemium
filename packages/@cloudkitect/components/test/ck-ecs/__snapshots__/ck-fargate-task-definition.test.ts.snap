// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Well Architected ECS Fargate Task Definition default fargate task definition 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "FGTaskDef8280256D": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "Cpu": "2048",
        "EphemeralStorage": {
          "SizeInGiB": 50,
        },
        "Family": "TestStackFGTaskDef8D72F2AF",
        "Memory": "4096",
        "NetworkMode": "awsvpc",
        "RequiresCompatibilities": [
          "FARGATE",
        ],
        "Tags": [
          {
            "Key": "CkComponents",
            "Value": "2.128.0",
          },
          {
            "Key": "CkVendor",
            "Value": "CloudKitect",
          },
        ],
        "TaskRoleArn": {
          "Fn::GetAtt": [
            "FGTaskDefTaskRoleAD31370C",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ECS::TaskDefinition",
      "UpdateReplacePolicy": "Retain",
    },
    "FGTaskDefTaskRoleAD31370C": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "ecs-tasks.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Tags": [
          {
            "Key": "CkComponents",
            "Value": "2.128.0",
          },
          {
            "Key": "CkVendor",
            "Value": "CloudKitect",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
